File: python/leetcode/173. Binary Search Tree Iterator.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class BSTIterator:
    def __init__(self, root: TreeNode):
        self.sorted = []
        self.pos = 0
        self.inorder(root)

    def next(self) -> int:
        """
        @return the next smallest number
        """
        val = self.sorted[self.pos]
        self.pos += 1
        return val

    def hasNext(self) -> bool:
        """
        @return whether we have a next smallest number
        """
        return self.pos < len(self.sorted)

    def inorder(self, root):
        if root is None:
            return

        self.inorder(root.left)
        self.sorted.append(root.val)
        self.inorder(root.right)


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()

File: go/leetcode/173. Binary Search Tree Iterator.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
type BSTIterator struct {
    sorted []int
    pos int
}
 

func Constructor(root *TreeNode) BSTIterator {
    bst := BSTIterator{[]int{}, 0}
    bst.inorder(root)     
    return bst
}


/** @return the next smallest number */
func (this *BSTIterator) Next() int {
    val := this.sorted[this.pos]
    this.pos++
    return val
}


/** @return whether we have a next smallest number */
func (this *BSTIterator) HasNext() bool {
    return this.pos < len(this.sorted)
}

func (bst *BSTIterator) inorder(root *TreeNode) {
    if root == nil {
        return 
    }
    
    bst.inorder(root.Left)
    bst.sorted = append(bst.sorted, root.Val)
    bst.inorder(root.Right)
}


/**
 * Your BSTIterator object will be instantiated and called as such:
 * obj := Constructor(root);
 * param_1 := obj.Next();
 * param_2 := obj.HasNext();
 */